src/playdate/scoreboards.nim

```diff
@@ -0,0 +1,175 @@
+import bindings/[api, types]
+import bindings/scoreboards
+
+# Only export public symbols, then import all
+export scoreboards
+{.hint[DuplicateModuleImport]: off.}
+import bindings/scoreboards {.all.}
+
+type 
+  PDScore* = object of RootObj
+    value*, rank*: uint32
+    player*: string
+
+  PDScoresList* = object of RootObj
+    boardID*: string
+    lastUpdated*: uint32
+    scores*: seq[PDScore]
+    # these properties are not implemented yet in the Playdate API
+    # playerIncluded*: uint32
+    # limit*: uint32
+
+  PDBoard* = object of RootObj
+    boardID*, name*: string
+
+  PDBoardsList* = object of RootObj
+    lastUpdated*: uint32
+    boards*: seq[PDBoard]
+
+  PersonalBestCallback* = proc(score: PDScore, errorMessage: string)
+  AddScoreCallback* = proc(score: PDScore, errorMessage: string)
+  BoardsListCallback* = proc(boards: PDBoardsList, errorMessage: string)
+  ScoresCallback* = proc(scores: PDScoresList, errorMessage: string)
+
+var
+  # The sdk callbacks unfortunately don't provide a userdata field to tag the callback with eg. the boardID
+  # So we need to keep track of the callbacks in the order they were called, which luckily is guaranteed by the Playdate API
+  # By inserting the callback at the start, it will be popped last: first in, first out
+  privatePersonalBestCallbacks = newSeq[PersonalBestCallback]()
+  privateAddScoreCallbacks = newSeq[AddScoreCallback]()
+  privateScoresCallbacks = newSeq[ScoresCallback]()
+  privateBoardsListCallbacks = newSeq[BoardsListCallback]()
+
+template invokeCallback(callbackSeqs, value, errorMessage, freeValue, builder, emptyValue: untyped) =
+  let callback = callbackSeqs.pop()
+  if value == nil and errorMessage == nil:
+      callback(emptyValue, "Playdate-nim: No value provided for callback")
+      return
+
+  if value == nil:
+    callback(emptyValue, $errorMessage)
+    return
+
+  try:
+    let nimObj = builder(value)
+    callback(nimObj, $errorMessage)
+  finally:
+    freeValue(value)
+
+
+proc newPDScore(value: uint32, rank: uint32, player: string): PDScore =
+  PDSCore(value: value, rank: rank, player: player)
+let emptyPDScore = newPDScore(value = 0, rank = 0, player = "")
+
+proc newPDScoresList(boardID: string, lastUpdated: uint32, scores: seq[PDScore]): PDScoresList =
+  PDScoresList(boardID: boardID, lastUpdated: lastUpdated, scores: scores)
+let emptyPDScoresList = newPDScoresList(boardID = "", lastUpdated = 0, scores = @[])
+
+proc newPDBoard(boardID: string, name: string): PDBoard =
+  PDBoard(boardID: boardID, name: name)
+
+proc newPDBoardsList(lastUpdated: uint32, boards: seq[PDBoard]): PDBoardsList =
+  PDBoardsList(lastUpdated: lastUpdated, boards: boards)
+let emptyPDBoardsList = newPDBoardsList(lastUpdated = 0, boards = @[])
+
+proc scoreBuilder(score: PDScoreRaw): PDScore =
+  newPDScore(
+    value = score.value.uint32,
+    rank = score.rank.uint32,
+    player = $score.player
+  )
+
+proc scoreBuilder(score: PDScorePtr): PDScore =
```

Comment:
Note sure about {.inline.} annotation for this. Some say the compiler is supposed to be smart enough to figure that out by itself.